crates/paimon/src/file_index/file_index_format.rs

```diff
@@ -0,0 +1,514 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashMap;
+
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+use crate::{
+    io::{FileIO, FileRead, InputFile, OutputFile},
+    Error,
+};
+
+pub const MAGIC: u64 = 1493475289347502;
+pub const EMPTY_INDEX_FLAG: i32 = -1;
+
+/// File index file format. All columns and offsets are stored in the header.
+///
+/// ```text
+///   _____________________________________    _____________________
+/// ｜     magic    ｜version｜head length ｜
+/// ｜-------------------------------------｜
+/// ｜            column number            ｜
+/// ｜-------------------------------------｜
+/// ｜   column 1        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜            HEADER
+/// ｜   column 2        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜  redundant length ｜redundant bytes ｜
+/// ｜-------------------------------------｜    ---------------------
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜             BODY
+/// ｜                BODY                 ｜
+/// ｜_____________________________________｜    _____________________
+///
+/// - `magic`: 8 bytes long
+/// - `version`: 4-byte integer
+/// - `head length`: 4-byte integer
+/// - `column number`: 4-byte integer
+/// - `column x`: variable-length UTF-8 string (length + bytes)
+/// - `index number`: 4-byte integer (number of index items below)
+/// - `index name x`: variable-length UTF-8 string
+/// - `start pos`: 4-byte integer
+/// - `length`: 4-byte integer
+/// - `redundant length`: 4-byte integer (for compatibility with future versions; content is zero in this version)
+/// - `redundant bytes`: variable-length bytes (for compatibility with future versions; empty in this version)
+/// - `BODY`: sequence of index data (concatenated index data for each column)
+/// ```
+#[derive(Debug)]
+pub struct FileIndexFormatWriter {
```

Comment:
I'm curious about the name. Why `FileIndexFormatWriter` instead of `FileIndexWriter`? Will we have other `FileIndexAbcWriter`?

Reply:
In Java semantics, FileIndexWriter functions as a specialized interface class, and Format was originally intended as a shell class to expose static methods for handling the Format structure, I made some adjustments to reflect this design.